plugins/android_boot/fu-android-boot-device.c

```diff
@@ -0,0 +1,441 @@
+			      GPtrArray *chunks,
+			      FuProgress *progress,
+			      GError **error)
+{
+	/* progress */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_set_steps(progress, chunks->len);
+
+	/* verify each chunk */
+	for (guint i = 0; i < chunks->len; i++) {
+		FuChunk *chk = g_ptr_array_index(chunks, i);
+		g_autofree guint8 *buf = g_malloc0(fu_chunk_get_data_sz(chk));
+		g_autoptr(GBytes) blob1 = fu_chunk_get_bytes(chk);
+		g_autoptr(GBytes) blob2 = NULL;
+
+		if (!fu_udev_device_pread(FU_UDEV_DEVICE(self),
+					  fu_chunk_get_address(chk),
+					  buf,
+					  fu_chunk_get_data_sz(chk),
+					  error)) {
+			g_prefix_error(error,
+				       "failed to read @0x%x: ",
+				       (guint)fu_chunk_get_address(chk));
+			return FALSE;
+		}
+		blob2 = g_bytes_new_static(buf, fu_chunk_get_data_sz(chk));
+		if (!fu_bytes_compare(blob1, blob2, error)) {
+			g_prefix_error(error,
+				       "failed to verify @0x%x: ",
+				       (guint)fu_chunk_get_address(chk));
+			return FALSE;
+		}
+		fu_progress_step_done(progress);
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_android_boot_device_write_firmware(FuDevice *device,
+				      FuFirmware *firmware,
+				      FuProgress *progress,
+				      FwupdInstallFlags flags,
+				      GError **error)
+{
+	FuAndroidBootDevice *self = FU_ANDROID_BOOT_DEVICE(device);
+	g_autoptr(GBytes) fw = NULL;
+	g_autoptr(GPtrArray) chunks = NULL;
+
+	/* get data to write */
+	fw = fu_firmware_get_bytes(firmware, error);
+	if (fw == NULL)
+		return FALSE;
+
+	if (g_getenv("FWUPD_ANDROID_BOOT_VERBOSE") != NULL)
+		fu_dump_bytes(G_LOG_DOMAIN, "write", fw);
+
+	if (g_bytes_get_size(fw) > fu_device_get_firmware_size_max(device)) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_INVALID_FILE,
+			    "firmware too large, got 0x%x, expected <= 0x%x",
+			    (guint)g_bytes_get_size(fw),
+			    (guint)fu_device_get_firmware_size_max(device));
+		return FALSE;
+	}
+	chunks = fu_chunk_array_new_from_bytes(fw, 0x0, 0x0, 10 * 1024);
+
+	/* progress */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_flag(progress, FU_PROGRESS_FLAG_GUESSED);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_ERASE, 72, NULL);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 20, NULL);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_VERIFY, 7, NULL);
+
+	/* erase */
+	g_debug("Erasing partition");
+	if (!fu_android_boot_device_erase(self, fu_progress_get_child(progress), error))
+		return FALSE;
+	fu_progress_step_done(progress);
+
+	/* write */
+	g_debug("Writing partition");
+	if (!fu_android_boot_device_write(self, chunks, fu_progress_get_child(progress), error))
+		return FALSE;
+	fu_progress_step_done(progress);
+
+	/* verify */
+	g_debug("Verifying partition");
+	if (!fu_android_boot_device_verify(self, chunks, fu_progress_get_child(progress), error))
+		return FALSE;
+	fu_progress_step_done(progress);
+
+	/* success */
+	return TRUE;
```

Comment:
I think you're too verbose in both your debug messages and comments.  If you need to show these messages for debugging then you should do prefixed errors

Reply:
Hmmm I can drop these as the progress logging already prints 'write', 'erase', and 'verify' in the logs.